### PR TITLE
[feature](function) support ip function named ipv4_cidr_to_range(addr, cidr)

### LINKED ISSUE #20584: [Enhancement] Add IP address process function for SQL like ClickHouse
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

ClickHouse support some [IP address process function](https://clickhouse.com/docs/zh/sql-reference/functions/ip-address-functions) for SQL. These are very useful for some network service cases.
Hope to add these function ASAP.Thanks.

### Solution

Add IP address process function for SQL like ClickHouse

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by nanfeng1999**
## Proposed changes

Issue Number: close #20584 

Accepts an IPv4 and a Int value containing the CIDR. Return a struct with two IPv4 containing the lower range(named min) and the higher range(named max) of the subnet.

Doris output:
image

### CODE REVIEW

**Comment by nanfeng1999** [comment]
run buildall

**Comment by nanfeng1999** [comment]
run buildall

**Comment by nanfeng1999** [comment]
run buildall

**Comment by sjyango** [comment]
run pipelinex_p0

**Comment by nanfeng1999** [comment]
run pipelinex_p0

**Comment by nanfeng1999** [comment]
run pipelinex_p0

**Comment by nanfeng1999** [comment]
run pipelinex_p0

**Comment by nanfeng1999** [comment]
run pipelinex_p0

**Comment by nanfeng1999** [comment]
run pipelinex_p0

**Comment by nanfeng1999** [comment]
run pipelinex_p0

**Comment by nanfeng1999** [comment]
run pipelinex_p0

**Comment by nanfeng1999** [comment]
run pipelinex_p0

File: gensrc/script/doris_builtins_functions.py [hunk 1]
```diff
@@ -2018,6 +2018,11 @@
         [['is_ipv6_string'], 'BOOLEAN', ['STRING'], 'ALWAYS_NULLABLE'],
         [['is_ip_address_in_range'], 'BOOLEAN', ['VARCHAR', 'VARCHAR'], 'ALWAYS_NOT_NULLABLE'],
         [['is_ip_address_in_range'], 'BOOLEAN', ['STRING', 'STRING'], 'ALWAYS_NOT_NULLABLE'],
+        [['ipv4_cidr_to_range'], 'ANY_STRUCT_TYPE', ['IPV4', 'TINYINT'], 'ALWAYS_NULLABLE'],
```
Comment (amorynan) [inline gensrc/script/doris_builtins_functions.py:2021]: why use ANY_STRUCT_TYPE?

File: be/src/vec/functions/function_ip.h [hunk 2]
```diff
@@ -832,4 +845,170 @@ class FunctionIsIPAddressInRange : public IFunction {
...
+        if (ip_column->is_nullable()) {
+            const auto* col_ip_nullable = check_and_get_column<ColumnNullable>(ip_column.get());
+            if (!col_ip_nullable) {
+                return Status::InvalidArgument("Illegal column {} of first argument of function {}",
+                                               ip_column->get_name(), get_name());
+            }
+            argument_nullmap[0] = col_ip_nullable->get_null_map_column_ptr();
+        }
+
+        if (cidr_column->is_nullable()) {
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:885]: same with IPV6 problems

File: be/src/vec/functions/function_ip.h [hunk 3]
```diff
@@ -832,4 +845,170 @@ class FunctionIsIPAddressInRange : public IFunction {
...
+            argument_nullmap[1] = col_cidr_nullable->get_null_map_column_ptr();
+        }
+
+        const auto* col_ip_column = check_and_get_column<ColumnType>(ip_column.get());
+        if (!col_ip_column) {
+            return Status::InvalidArgument("Illegal column {} of first argument of function {}",
+                                           ip_column->get_name(), get_name());
+        }
+
+        const auto* col_cidr_column = check_and_get_column<ColumnVector<Int8>>(cidr_column.get());
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:901]: use data type to check

File: be/src/vec/functions/function_ip.h [hunk 4]
```diff
@@ -832,4 +845,170 @@ class FunctionIsIPAddressInRange : public IFunction {
...
+
+        return make_nullable(std::make_shared<DataTypeStruct>(dataTypes, names));
+    }
+
+    bool use_default_implementation_for_nulls() const override { return true; }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) const override {
+        ColumnWithTypeAndName& cidr_column = block.get_by_position(arguments[1]);
+        if (cidr_column.type->get_type_id() != TypeIndex::Int8) {
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:982]: cidr [0 - 255] ? why int8?

File: be/src/vec/functions/function_ip.h [hunk 5]
```diff
@@ -832,4 +845,170 @@ class FunctionIsIPAddressInRange : public IFunction {
...
+                    "Illegal type {} of second argument of function {}, expected Int8",
+                    cidr_column.name, get_name());
+        }
+
+        ColumnWithTypeAndName& ip_column = block.get_by_position(arguments[0]);
+        ColumnsWithTypeAndName input_column {ip_column, cidr_column};
+
+        // RPC does not support uint type.
+        // Passed through RPC may include IPv4, int8, int16, int32, and int64.
+        switch (ip_column.type->get_type_id()) {
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:993]: I think here is no need to switch case

**Comment by nanfeng1999** [comment]
run buildall

File: be/src/vec/functions/function_ip.h [hunk 6]
```diff
@@ -741,6 +752,129 @@ class FunctionIsIPAddressInRange : public IFunction {
...
+
+        return std::make_shared<DataTypeStruct>(DataTypes {element, element},
+                                                Strings {"min", "max"});
+    }
+
+    Status execute_impl(FunctionContext* context, Block& block, const ColumnNumbers& arguments,
+                        size_t result, size_t input_rows_count) const override {
+        ColumnWithTypeAndName& cidr_column = block.get_by_position(arguments[1]);
+        // Use int16 to represent cidr to prevent cidr value from overflowing
+        if (cidr_column.type->get_type_id() != TypeIndex::Int16) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_ip.h:790]: check the type in `get_return_type_impl`, no need check again

File: be/src/vec/functions/function_ip.h [hunk 7]
```diff
@@ -741,6 +752,129 @@ class FunctionIsIPAddressInRange : public IFunction {
...
+    static constexpr auto name = "ipv4_cidr_to_range";
+    static FunctionPtr create() { return std::make_shared<FunctionIPv4CIDRToRange>(); }
+
+    String get_name() const override { return name; }
+
+    size_t get_number_of_arguments() const override { return 2; }
+
+    DataTypePtr get_return_type_impl(const DataTypes& arguments) const override {
+        WhichDataType first_arg_type = arguments[0];
+        if (!(first_arg_type.is_ipv4() || first_arg_type.is_native_int())) {
```
Comment (HappenLee) [inline be/src/vec/functions/function_ip.h:766]: why here need support other int type, seems wired?

File: be/src/vec/functions/function_ip.h [hunk 8]
```diff
@@ -741,6 +752,129 @@ class FunctionIsIPAddressInRange : public IFunction {
...
+            auto ip = vec_ip_input[i];
+            auto cidr = vec_cidr_input[i];
+            if (IPV4_MIN_NUM_VALUE <= ip && ip <= IPV4_MAX_NUM_VALUE && 0 <= cidr &&
+                cidr <= max_cidr_mask) {
+                auto range = apply_cidr_mask(ip, cidr);
+                vec_lower_range_output[i] = range.first;
+                vec_upper_range_output[i] = range.second;
+            } else {
+                return Status::RuntimeError("Invalid row {}, IP or cidr is out of range", i);
+            }
```
Comment (HappenLee) [inline be/src/vec/functions/function_ip.h:853]: if only support ipv4 type, maybe we do not need check ip is valied?

**Comment by nanfeng1999** [comment]
run buildall

**Comment by nanfeng1999** [comment]
run buildall

**Comment by nanfeng1999** [comment]
run buildall

**Comment by nanfeng1999** [comment]
run buildall

**Comment by nanfeng1999** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by zclllyybb — APPROVED** [review]

**Review by wangbo — APPROVED** [review]
LGTM

**Pull request merged**